### PR TITLE
FIX: Fix ReceptiveField edge handling

### LINKED ISSUE #4253: Edge artefacts in ReceptiveField
I chatted briefly with @choldgraf about this issue today, but I've been giving the new `ReceptiveField` code a try on some old data today. And I'm getting strange artefacts on the reconstructed signal. 

Here I've been using the stimulus reconstruction approach (backward model) to predict my stimulus from the EEG. The data are 10s epochs, and I have about 50-70 of those.

Here's what the reconstruction looks like (all the reconstructed epochs are concatenated inside the `predict` method). X axis is time samples @ 256Hz.
figure_3
figure_2
figure_1

The artefact is located in the first `len(time_lags)` samples of each epoch, so I assume it has to do with how the delays are handled in `predict`.

### PR DESCRIPTION
**Pull request by Eric89GXL**
This fixes the edge handling when using Ridge, and it also changes the API a bit.

I also put in a little API change up for discussion. The output of `predict` is now `(n_times[, n_epochs], n_outputs)` instead of what's in `master` which is `(n_times * n_epochs[, n_outputs])`. This is really the result of two changes/ideas:

1. If you pass in something with `n_epochs` dimension, you now get something back with an `n_epochs` dimension.

2. You always have the `n_outputs` dimension, even if it's a singleton. In our definition of `fit` we have `y : array, shape (n_times[, n_epochs], n_outputs)`, i.e. the output dimension is always there, so it seems like we should always have it in `predict`, too.

Not sure if this violates some standard `sklearn` terms or not. Maybe we need to add the `vectorize=True` argument eventually to make the epochs stuff `sklearn`-compatible.

Closes #4253.

### CODE REVIEW

**Comment by Eric89GXL** [comment]
The example ([old](https://mne-tools.github.io/dev/auto_examples/decoding/plot_receptive_field.html) and [new](https://4778-1301584-gh.circle-artifacts.com/0/tmp/circle-artifacts.ACvXO3C/html/auto_examples/decoding/plot_receptive_field.html)) and tutorial ([old](https://mne-tools.github.io/dev/auto_tutorials/plot_receptive_field.html) and [new](https://4778-1301584-gh.circle-artifacts.com/0/tmp/circle-artifacts.ACvXO3C/html/auto_tutorials/plot_receptive_field.html)) show almost no changes, which is probably expected here.

**Comment by Eric89GXL** [comment]
Failures seem unrelated to this PR

**Comment by agramfort** [comment]
i'll let @nbara and @choldgraf check and give their +1 here.

**Comment by nbara** [comment]
Working fine on my data now so +1 for me

**Comment by agramfort** [comment]
merge when you guys are happy

File: mne/decoding/receptive_field.py [hunk 1]
```diff
@@ -56,10 +57,15 @@ class ReceptiveField(BaseEstimator):
     ``delays_``: array, shape (n_delays,), dtype int
         The delays used to fit the model, in indices. To return the delays
         in seconds, use ``self.delays_ / self.sfreq``
-    ``keep_samples_`` : slice
+    ``valid_samples_`` : slice
         The rows to keep during model fitting after removing rows with
-        missing values due to time delaying.
+        missing values due to time delaying. This can be used to get an
+        output equivalent to that of ``mode='valid'`` of
```
Comment (choldgraf) [inline mne/decoding/receptive_field.py]: Feels a bit awkward, how about "This can be used to get an output equivalent to using :func:`numpy.convolve` or :func:`numpy.correlate` with ``mode='valid'``."?

File: mne/decoding/receptive_field.py [hunk 2]
```diff
@@ -203,14 +223,25 @@ def predict(self, X):
 
         Returns
         -------
-        y_pred : array, shape (n_times * n_epochs[, n_outputs])
-            The output predictions with time concatenated.
+        y_pred : array, shape (n_times[, n_epochs], n_outputs)
+            The output predictions. Note that valid samples can be
```
Comment (choldgraf) [inline mne/decoding/receptive_field.py]: I think we should be more explicit about what we mean when we say "valid". Maybe we can just add in a short bit like this: "Note that valid samples (those unaffected by edge artifacts during the time delaying step) can be obtained...".

File: mne/decoding/receptive_field.py [hunk 3]
```diff
@@ -240,14 +271,13 @@ def score(self, X, y):
         X, y = self._check_dimensions(X, y, predict=True)
         n_times, n_epochs, n_outputs = y.shape
         y_pred = self.predict(X)
-
-        y_pred = y_pred.reshape(y.shape, order='F')
-        y_pred = y_pred[self.keep_samples_]
-        y = y[self.keep_samples_]
+        y_pred = y_pred[self.valid_samples_]
```
Comment (choldgraf) [inline mne/decoding/receptive_field.py]: I think this is where it might break sklearn compatibility. I could be wrong though, since we do the reshaping below

File: mne/decoding/receptive_field.py [hunk 4]
```diff
@@ -294,7 +324,33 @@ def _check_dimensions(self, X, y, predict=False):
         return X, y
 
 
-def _delay_time_series(X, tmin, tmax, sfreq, newaxis=0, axis=0):
+def _pad_time_series(X, n_delays, fill_mean=True):
```
Comment (choldgraf) [inline mne/decoding/receptive_field.py]: don't we have some function like this already implemented to do the zero padding for things like FFT? (e.g., zero padding up to a length of 2**N). Just asking for the sake of DRY :)

Comment (Eric89GXL) [inline mne/decoding/receptive_field.py]: This one is a bit different because it isn't trying to get to a next fast FFT length, just trying to prevent overlapping during convolution. So I don't think so.

File: mne/decoding/receptive_field.py [hunk 5]
```diff
@@ -311,12 +367,15 @@ def _delay_time_series(X, tmin, tmax, sfreq, newaxis=0, axis=0):
     newaxis : int
         The axis in the output array that corresponds to time delays.
         Defaults to 0, for the first axis.
+    fill_mean : bool
```
Comment (choldgraf) [inline mne/decoding/receptive_field.py]: If True...the fill value will be the mean along the time dimension of the feature. If False, the fill value will be zero.

File: mne/decoding/receptive_field.py (hunk 3 above)
Comment (Eric89GXL) [inline mne/decoding/receptive_field.py]: To what extent do we need `sklearn` compatibility?

Maybe it would be good to come up with some small example that shows how to use it in a `sklearn` chain (in a separate issue/PR probably). That way we can think properly about how to do it?

File: mne/decoding/tests/test_receptive_field.py [hunk 6]
```diff
@@ -34,6 +34,39 @@
 picks = picks[:2]
 
 
+@requires_sklearn_0_15
+def test_rank_deficiency():
```
Comment (choldgraf) [inline mne/decoding/tests/test_receptive_field.py]: nice

**Comment by choldgraf** [comment]
Some small explanatory points in the review, but other than that LGTM. Thanks for digging into this and nice catch w/ the rank deficiency stuff!

**Comment by choldgraf** [comment]
also pinging @kingjr on this thread because he may have some good insights about what does / doesn't == sklearn compatibility.

File: mne/decoding/receptive_field.py (hunk 3 above)
Comment (Eric89GXL) [inline mne/decoding/receptive_field.py]: This is also within `score`, so I'm not sure how/why this would break `sklearn` compatibility.

**Comment by choldgraf** [comment]
LGTM assuming @kingjr doesn't have strong opinions about the sklearn compatibility stuff. I think this could be fixed with a `vectorize=True` parameter that we figure out in the future so my intuition is that it's OK

**Comment by agramfort** [comment]
does the code crash if y.shape = (n_times,) and not y.shape = (n_times, 1).

in sklearn convention if you pass (n_times, 1) then you do est.coef_[0] but if (n_times,) then est.coef_ is what you need.

if current code matches this API +1 for MRG

**Comment by kingjr** [comment]
No strong opinion ATM, but I haven't use this yet

**Comment by Eric89GXL** [comment]
> does the code crash if y.shape = (n_times,) and not y.shape = (n_times, 1).

This is now tested and all is good, including the convention about 1D vs 2D with shape (N,1).

Ready for merge from my end. @choldgraf can you take one last look, probably even looking at the docstring changes (about dimensions) should be enough to show the new stuff.

File: mne/decoding/receptive_field.py [hunk 7]
```diff
@@ -39,27 +39,33 @@ class ReceptiveField(BaseEstimator):
         float is passed, it will be interpreted as the `alpha` parameter
         to be passed to a Ridge regression model. If `None`, then a Ridge
         regression model with an alpha of 0 will be used.
-    fit_intercept : bool
+    fit_intercept : bool | None
         If True (default), the sample mean is removed before fitting.
-        Ignored if ``estimator`` is a :class:`sklearn.base.BaseEstimator`.
+        If ``estimator`` is a :class:`sklearn.base.BaseEstimator`,
+        this must be None or match ``estimator.fit_intercept``.
```
Comment (choldgraf) [inline mne/decoding/receptive_field.py]: In this case why not enforce it to be None and then just pull from `self.estimator.fit_intercept`?

File: mne/decoding/receptive_field.py [hunk 8]
```diff
@@ -39,27 +39,33 @@ class ReceptiveField(BaseEstimator):
...
         coefficient).
 
     Attributes
     ----------
-    ``coef_`` : array, shape (n_outputs, n_features, n_delays)
+    ``coef_`` : array, shape ([n_outputs, ]n_features, n_delays)
         The coefficients from the model fit, reshaped for easy visualization.
-        If you want the raw (1d) coefficients, access them from the estimator
-        stored in ``self.estimator_``.
+        If ``y`` during :meth:`mne.decoding.ReceptiveField.fit` has one
```
Comment (choldgraf) [inline mne/decoding/receptive_field.py]: I'd say "in the call to :meth:etcetc, if ``y`` has one dimension..."

keeps things less broken-up semantically

File: mne/decoding/receptive_field.py [hunk 9]
```diff
@@ -149,20 +152,30 @@ def fit(self, X, y):
...
 
         if isinstance(self.estimator, numbers.Real):
+            if self.fit_intercept is None:
+                self.fit_intercept = True
             estimator = TimeDelayingRidge(self.tmin, self.tmax, self.sfreq,
                                           alpha=self.estimator,
                                           fit_intercept=self.fit_intercept)
         elif is_regressor(self.estimator):
             estimator = clone(self.estimator)
+            if self.fit_intercept is not None and \
```
Comment (choldgraf) [inline mne/decoding/receptive_field.py]: so here we could just do `if self.fit_intercept is not None: "raise the error"` yeah?

**Comment by Eric89GXL** [comment]
I'd rather tackle those two other suggestions in another PR -- they seem mostly orthogonal and I'm not sure when I can get to it fully.

File: mne/decoding/receptive_field.py (hunk 7 above)
Comment (Eric89GXL) [inline mne/decoding/receptive_field.py]: I don't mind if the person does `fit_intercept=True` in both places (or `fit_intercept=False` in both places), i.e. they just need to match

Comment (Eric89GXL) [inline mne/decoding/receptive_field.py]: ... unless you think there is some risk of ambiguity or it's a bad API?

**Comment by choldgraf** [comment]
+1 LGTM

**Comment by Eric89GXL** [comment]
Thanks for the reviews. @agramfort is happy enough, too, so in it goes!
